File: src/systems/CombatSystem.js
import { EventBus } from '../core/EventBus.js';
import { BUILDINGS } from '../data/BuildingConfig.js';
import { Projectile } from '../entities/Projectile.js';
import { PHASE } from '../core/Constants.js';

export class CombatSystem {
  constructor(gameState) {
    this.state = gameState;
  }

  update(dt) {
    if (this.state.phase !== PHASE.BATTLE) return;

    this._updateTowers(dt);
    this._updateTroops(dt);
    this._updateEnemies(dt);
    this._updateProjectiles(dt);
    this._cleanupDead();
    this._updateEffects(dt);
  }

  _updateTowers(dt) {
    for (const building of this.state.buildings) {
      if (building.configId !== 'ARCHER_TOWER') continue;
      if (building.isBuilding) continue;

      building.attackCooldown -= dt;
      if (building.attackCooldown > 0) continue;

      const range = building.getStat('attackRange');
      const damage = building.getStat('attackDamage');
      const attackSpeed = building.getStat('attackSpeed');

      if (!range || !damage) continue;

      const config = building.getConfig();
      const towerCol = building.col + config.tileWidth / 2;
      const towerRow = building.row + config.tileHeight / 2;

      let nearest = null;
      let nearestDist = Infinity;

      for (const enemy of this.state.enemies) {
        if (enemy.state === 'DEAD') continue;
        const dist = Math.hypot(enemy.col - towerCol, enemy.row - towerRow);
        if (dist <= range && dist < nearestDist) {
          nearest = enemy;
          nearestDist = dist;
        }
      }

      if (nearest) {
        const proj = new Projectile(towerCol, towerRow, nearest, damage);
        this.state.projectiles.push(proj);
        building.attackCooldown = attackSpeed;
      }
    }
  }

  // Stationary troops - only attack enemies within range, never chase
  _updateTroops(dt) {
    for (const troop of this.state.troops) {
      if (troop.state === 'DEAD') continue;

      // Find nearest enemy within range
      let nearest = null;
      let nearestDist = Infinity;

      for (const enemy of this.state.enemies) {
        if (enemy.state === 'DEAD') continue;
        const dist = troop.distanceTo(enemy.col, enemy.row);
        if (dist <= troop.range && dist < nearestDist) {
          nearest = enemy;
          nearestDist = dist;
        }
      }

      if (!nearest) {
        troop.state = 'IDLE';
        troop.target = null;
        continue;
      }

      troop.target = nearest;
      troop.state = 'ATTACKING';
      troop.attackCooldown -= dt;

      if (troop.attackCooldown <= 0) {
        if (troop.range > 2) {
          const proj = new Projectile(troop.col, troop.row, nearest, troop.damage);
          this.state.projectiles.push(proj);
        } else {
          nearest.takeDamage(troop.damage);
          this.state.addEffect('damage', nearest.col, nearest.row, troop.damage);
        }
        troop.attackCooldown = troop.attackSpeed;
      }
    }
  }

  // Cats chase nearest troop and attack. No HQ damage — wave fail handled by WaveSystem.
  _updateEnemies(dt) {
    for (const enemy of this.state.enemies) {
      if (enemy.state === 'DEAD') continue;

      // Find nearest alive troop globally (not just in range)
      let nearest = null;
      let nearestDist = Infinity;

      for (const troop of this.state.troops) {
        if (troop.state === 'DEAD') continue;
        const dist = enemy.distanceTo(troop.col, troop.row);
        if (dist < nearestDist) {
          nearest = troop;
          nearestDist = dist;
        }
      }

      if (!nearest) {
        // No troops left — wander toward base center (they'll be cleaned up by WaveSystem failure logic)
        enemy.state = 'MOVING';
        const cx = 15, cy = 15;
        const dx = cx - enemy.col;
        const dy = cy - enemy.row;
        const dist = Math.hypot(dx, dy);
        if (dist > 0.01) {
          enemy.col += (dx / dist) * enemy.speed * dt;
          enemy.row += (dy / dist) * enemy.speed * dt;
        }
        continue;
      }

      if (nearestDist <= enemy.range) {
        // Attack
        enemy.state = 'ATTACKING';
        enemy.attackCooldown -= dt;
        if (enemy.attackCooldown <= 0) {
          nearest.takeDamage(enemy.damage);
          this.state.addEffect('damage', nearest.col, nearest.row, enemy.damage);
          enemy.attackCooldown = enemy.attackSpeed;
        }
      } else {
        // Chase nearest troop
        enemy.state = 'MOVING';
        const dx = nearest.col - enemy.col;
        const dy = nearest.row - enemy.row;
        const dist = Math.hypot(dx, dy);
        if (dist > 0) {
          enemy.col += (dx / dist) * enemy.speed * dt;
          enemy.row += (dy / dist) * enemy.speed * dt;
        }
      }
    }
  }

  _updateProjectiles(dt) {
    for (const proj of this.state.projectiles) {
      if (!proj.alive) continue;

      const target = proj.targetEntity;
      if (!target || target.state === 'DEAD' || target.hp <= 0) {
        proj.alive = false;
        continue;
      }

      proj.prevCol = proj.col;
      proj.prevRow = proj.row;

      const dx = target.col - proj.col;
      const dy = target.row - proj.row;
      const dist = Math.hypot(dx, dy);

      if (dist < 0.3) {
        target.takeDamage(proj.damage);
        this.state.addEffect('damage', target.col, target.row, proj.damage);
        proj.alive = false;
      } else {
        proj.col += (dx / dist) * proj.speed * dt;
        proj.row += (dy / dist) * proj.speed * dt;

        proj.flightProgress += dt * proj.speed / dist;
        proj.arcHeight = Math.sin(proj.flightProgress * Math.PI) * 15;
      }
    }
  }

  _cleanupDead() {
    for (let i = this.state.enemies.length - 1; i >= 0; i--) {
      const enemy = this.state.enemies[i];
      if (enemy.state === 'DEAD') {
        this.state.addResource('water', enemy.reward.water);
        this.state.addResource('milk', enemy.reward.milk);
        this.state.addXP(enemy.xp);
        this.state.addEffect('reward', enemy.col, enemy.row, `+${enemy.reward.water}W`);
        this.state.enemies.splice(i, 1);
      }
    }

    for (let i = this.state.troops.length - 1; i >= 0; i--) {
      if (this.state.troops[i].state === 'DEAD') {
        this.state.troops.splice(i, 1);
      }
    }

    this.state.projectiles = this.state.projectiles.filter(p => p.alive);
  }

  _updateEffects(dt) {
    for (let i = this.state.effects.length - 1; i >= 0; i--) {
      const effect = this.state.effects[i];
      effect.progress += dt / effect.duration;
      if (effect.progress >= 1) {
        this.state.effects.splice(i, 1);
      }
    }
  }
}

File: ios/BiteDefense/Systems/CombatSystem.swift
import Foundation

/// Minimal combat loop — troops attack nearest enemy in range, enemies chase
/// nearest troop (or HQ if none), attacks resolve on cooldown. Direct-line
/// movement (no pathfinding; matches JS reference). M10+ can upgrade to
/// path-based routing if needed.
final class CombatSystem {
    private unowned let state: GameState

    init(state: GameState) {
        self.state = state
    }

    func update(dt: Double) {
        guard state.phase == .battle else { return }
        updateTroops(dt: dt)
        updateEnemies(dt: dt)
        cleanupDead()
    }

    // MARK: - Troops

    private func updateTroops(dt: Double) {
        for i in state.troops.indices {
            guard !state.troops[i].isDead,
                  state.troops[i].state != .garrisoned else { continue }

            let t = state.troops[i]
            let range = t.def.range(level: t.level)

            // Find nearest alive enemy within range.
            var nearestIdx: Int? = nil
            var nearestDist: Double = .infinity
            for j in state.enemies.indices {
                guard !state.enemies[j].isDead else { continue }
                let d = hypot(state.enemies[j].col - t.col, state.enemies[j].row - t.row)
                if d <= range && d < nearestDist {
                    nearestIdx = j
                    nearestDist = d
                }
            }

            guard let idx = nearestIdx else {
                state.troops[i].state = .idle
                continue
            }
            state.troops[i].state = .attacking
            state.troops[i].attackCooldown -= dt
            if state.troops[i].attackCooldown <= 0 {
                let damage = t.def.damage(level: t.level)
                let isRanged = t.def.range(level: t.level) > 2
                let tx = t.col, ty = t.row
                let ex = state.enemies[idx].col, ey = state.enemies[idx].row
                if isRanged {
                    EventBus.shared.send(.projectileFired(fromCol: tx, fromRow: ty,
                                                           toCol: ex, toRow: ey,
                                                           damage: damage))
                }
                state.enemies[idx].hp -= damage
                EventBus.shared.send(.enemyDamaged(enemyId: state.enemies[idx].id,
                                                   amount: damage,
                                                   col: ex, row: ey))
                if state.enemies[idx].hp <= 0 {
                    state.enemies[idx].state = .dead
                    EventBus.shared.send(.enemyDied(enemyId: state.enemies[idx].id))
                }
                state.troops[i].attackCooldown = t.def.attackSpeed(level: t.level)
            }
        }
    }

    // MARK: - Enemies

    private func updateEnemies(dt: Double) {
        let hqIdx = state.buildings.firstIndex(where: { $0.type == .dogHQ })

        for i in state.enemies.indices {
            guard !state.enemies[i].isDead else { continue }

            let def = state.enemies[i].def

            // Find nearest alive battlefield troop (not garrisoned).
            var nearestTroopIdx: Int? = nil
            var nearestDist: Double = .infinity
            for j in state.troops.indices {
                let t = state.troops[j]
                guard !t.isDead, t.state != .garrisoned else { continue }
                let d = hypot(t.col - state.enemies[i].col, t.row - state.enemies[i].row)
                if d < nearestDist {
                    nearestTroopIdx = j
                    nearestDist = d
                }
            }

            if let j = nearestTroopIdx {
                if nearestDist <= def.range {
                    state.enemies[i].state = .attacking
                    state.enemies[i].attackCooldown -= dt
                    if state.enemies[i].attackCooldown <= 0 {
                        state.troops[j].hp -= state.enemies[i].damage
                        let tx = state.troops[j].col
                        let ty = state.troops[j].row
                        EventBus.shared.send(.troopDamaged(troopId: state.troops[j].id,
                                                             amount: state.enemies[i].damage,
                                                             col: tx, row: ty))
                        if state.troops[j].hp <= 0 {
                            state.troops[j].state = .dead
                            EventBus.shared.send(.troopDied(troopId: state.troops[j].id))
                        }
                        state.enemies[i].attackCooldown = def.attackSpeed
                    }
                } else {
                    moveEnemy(i, toX: state.troops[j].col, toY: state.troops[j].row, dt: dt)
                }
                continue
            }

            // No battlefield troops — attack HQ.
            if let hi = hqIdx, state.buildings[hi].hp > 0 {
                let hq = state.buildings[hi]
                let cx = Double(hq.col) + Double(hq.def.tileWidth) / 2
                let cy = Double(hq.row) + Double(hq.def.tileHeight) / 2
                let distToHQ = hypot(cx - state.enemies[i].col, cy - state.enemies[i].row)
                if distToHQ <= def.range + 1 {
                    state.enemies[i].state = .attacking
                    state.enemies[i].attackCooldown -= dt
                    if state.enemies[i].attackCooldown <= 0 {
                        var updated = hq
                        updated.hp = max(0, updated.hp - state.enemies[i].damage)
                        state.buildings[hi] = updated
                        EventBus.shared.send(.buildingDamaged(buildingId: hq.id,
                                                                hp: updated.hp,
                                                                maxHP: updated.maxHP,
                                                                amount: state.enemies[i].damage))
                        state.enemies[i].attackCooldown = def.attackSpeed
                    }
                } else {
                    moveEnemy(i, toX: cx, toY: cy, dt: dt)
                }
            }
        }
    }

    private func moveEnemy(_ i: Int, toX: Double, toY: Double, dt: Double) {
        state.enemies[i].state = .moving
        let def = state.enemies[i].def
        let dx = toX - state.enemies[i].col
        let dy = toY - state.enemies[i].row
        let d = hypot(dx, dy)
        guard d > 0 else { return }
        state.enemies[i].col += dx / d * def.speed * dt
        state.enemies[i].row += dy / d * def.speed * dt
    }

    // MARK: - Cleanup

    private func cleanupDead() {
        // Reward for dead enemies, then remove.
        var i = state.enemies.count - 1
        while i >= 0 {
            if state.enemies[i].isDead {
                let e = state.enemies[i]
                state.add(e.def.rewardWater, to: .water)
                state.add(e.def.rewardMilk, to: .milk)
                state.addXP(e.def.xp)
                state.enemies.remove(at: i)
            }
            i -= 1
        }
        // Remove dead troops visually — keep them in state until wave end so
        // the WaveSystem can detect "all troops dead". Actually we remove dead
        // troops here and let WaveSystem use troop.state==.dead check instead.
        // We keep them, the isDead check handles it.
    }
}
